File: rocket_sim.py
import math

# --------------------
# Rocket body aerodynamics (6-inch diameter)
# --------------------
BODY_CD = 0.4
BODY_DIAMETER = 0.1524                              # 6 inches in metres
BODY_AREA = math.pi * (BODY_DIAMETER / 2) ** 2     # ~0.01824 m²


def rocket_sim(x, v, tilt_deg, airbrake, ground_pressure, verbose=False):
    # --------------------
    # Constants
    # --------------------
    dt = 0.01          # time step (s)
    g = 9.80665        # gravity (m/s^2)
    R = 287.05         # J/(kg*K)
    L = 0.0065         # K/m
    mass = 113.0       # kg

    time = 0.0

    # --------------------
    # Ground conditions
    # --------------------
    T0 = 288.15        # K
    P0_pa = ground_pressure

    # --------------------
    # Helper functions
    # --------------------
    def air_density(h):
        """ISA air density — same formula as controller.py."""
        T = max(T0 - L * h, 1.0)
        P = P0_pa * (T / T0) ** (g / (R * L))
        return max(P / (R * T), 0.001)

    def airbrake_coeff_area(deploy):
        """Airbrake contribution only (body drag is added separately)."""
        Cd = 0.4
        A = 0.001848 + (0.021935 - 0.001848) * deploy  # 2.86479 in² to 34 in²
        return Cd, A

    # --------------------
    # Simulation loop
    # --------------------
    cos_tilt = max(math.cos(math.radians(abs(tilt_deg))), 1e-6)

    while True:
        # --- Air density at current altitude ---
        rho = air_density(x)

        # --- Airspeed along rocket axis ---
        v_air = abs(v) / cos_tilt

        # --- Body drag ---
        Fd_body = 0.5 * rho * v_air**2 * BODY_CD * BODY_AREA

        # --- Airbrake drag ---
        Cd_brake, A_brake = airbrake_coeff_area(airbrake)
        Fd_brake = 0.5 * rho * v_air**2 * Cd_brake * A_brake

        # --- Total drag (always opposes upward motion during coast) ---
        Fd = Fd_body + Fd_brake

        # --- Acceleration ---
        a = -g - Fd / mass

        # --- Kinematics ---
        v_next = v + a * dt
        x_next = x + v * dt + 0.5 * a * dt**2

        # --- Apogee condition ---
        if v_next < 0:
            if verbose:
                print(f"Apogee reached at {x:.2f} m after {time:.2f} s")
            break

        # --- Update ---
        v = v_next
        x = x_next
        time += dt
    return x_next


# --------------------
# CLI entry point
# --------------------
if __name__ == "__main__":
    def get_float(prompt, default):
        raw = input(f"{prompt} [default {default}]: ").strip()
        return float(raw) if raw else float(default)

    print("=== Rocket Coast Simulation ===")
    print("Press Enter to accept the default value.\n")

    while True:
        x0            = get_float("Initial altitude (m)",            0.0)
        v0            = get_float("Initial vertical velocity (m/s)", 200.0)
        tilt_deg      = get_float("Tilt off-axis (deg)",             0.0)
        airbrake      = get_float("Airbrake deployment level (0-1)", 0.0)
        ground_p_pa   = get_float("Ground pressure (Pa)",            101325.0)

        apogee = rocket_sim(x0, v0, tilt_deg, airbrake, ground_p_pa, verbose=True)
        print(f"  → Predicted apogee: {apogee:.2f} m\n")

        again = input("Run another simulation? (y/n) [n]: ").strip().lower()
        if again != "y":
            break

File: controller_in_rust/src/rocket_sim.rs
// Rocket body aerodynamics (6-inch diameter)
const BODY_CD: f64 = 0.4;
const BODY_DIAMETER: f64 = 0.1524;                                    // 6 inches in metres
const BODY_AREA: f64 = std::f64::consts::PI * (BODY_DIAMETER / 2.0) * (BODY_DIAMETER / 2.0); // ~0.01824 m²

/// Simulates coast phase from given conditions and returns predicted apogee (m).
pub fn rocket_sim(x: f64, v: f64, tilt_deg: f64, airbrake: f64, ground_pressure: f64) -> f64 {
    const DT: f64 = 0.01;
    const G: f64 = 9.80665;
    const R: f64 = 287.05;
    const L: f64 = 0.0065;
    const MASS: f64 = 113.0;
    const T0: f64 = 288.15;
    let p0_pa = ground_pressure;

    // ISA air density — same formula as controller.rs
    let air_density = |h: f64| -> f64 {
        let t = (T0 - L * h).max(1.0);
        let p = p0_pa * (t / T0).powf(G / (R * L));
        (p / (R * t)).max(0.001)
    };

    // Airbrake contribution only (body drag is added separately)
    let airbrake_coeff_area = |deploy: f64| -> (f64, f64) {
        (0.4, 0.001848 + (0.021935 - 0.001848) * deploy) // 2.86479 in² to 34 in²
    };

    let cos_tilt = tilt_deg.abs().to_radians().cos().max(1e-6);

    let mut x = x;
    let mut v = v;

    loop {
        // Air density at current altitude
        let rho = air_density(x);

        // Airspeed along rocket axis (vertical velocity is one leg, axial is hypotenuse)
        let v_air = v.abs() / cos_tilt;
        let dynamic_pressure = 0.5 * rho * v_air * v_air;

        // Body drag
        let fd_body = dynamic_pressure * BODY_CD * BODY_AREA;

        // Airbrake drag
        let (cd_brake, a_brake) = airbrake_coeff_area(airbrake);
        let fd_brake = dynamic_pressure * cd_brake * a_brake;

        // Total drag (always opposes upward motion during coast)
        let fd = fd_body + fd_brake;

        let acc = -G - fd / MASS;
        let v_next = v + acc * DT;
        let x_next = x + v * DT + 0.5 * acc * DT * DT;

        if v_next < 0.0 {
            return x_next;
        }

        v = v_next;
        x = x_next;
    }
}
